xls/contrib/mlir/IR/xls_ops.td

```diff
@@ -1226,6 +1255,55 @@ def Xls_ConstantScalarOp : Xls_Op<"constant_scalar", [ConstantLike, Pure]> {
   let hasFolder = 1;
 }
 
+def Xls_LiteralOp : Xls_Op<"literal", [
+  ConstantLike,
+  Pure,
+  IsolatedFromAbove,
+  SingleBlockImplicitTerminator<"YieldOp">
+]> {
+  let summary = "Complex Literal";
+  let description = [{
+    Produces a complex (non-bits) constant value from a region containing
+    simple constants or tuple/array construction operations.
+
+    The body must be terminated by a `xls.yield` op, which returns the
+    constructed value.
+
+    Besides this yield, only the following operations are permissible inside
```

Comment:
I think it would be computationally cheaper to just have a helper function with the TypeIDs of these ops rather than using traits (you'd be doing pointer comparison based on name rather than doing a linear scan per op through trait list). If these are this well known and we don't need to support any other dialects/unknown supported op types.

Reply:
Not an unreasonable suggestion, but personally I like that this is all having tablegen as the source of truth, and I'm not majorly worried about linear scans of literal regions in the verifier (they'll be small).